### PR TITLE
Refactor eigendecomposition-related matrix functions

### PR DESCRIPTION
**Pull request by runame**
Refactors how matrix functions related to eigendecompositions are structured by introducing a new `matrix_eigendecomposition` function which allows computing eigendecompositions with `torch.linalg.eigh` or the QR algorithm.

This prepares support for regular Shampoo with warm-started QR algorithm.

### CODE REVIEW

File: matrix_functions.py [hunk 1]
```diff
@@ -206,13 +263,85 @@ def matrix_eigenvalue_decomposition(
     return L.to(device=current_device), Q.to(device=current_device)
 
 
+def _approximate_eigenvalues_criterion_below_or_equal_tolerance(
```
[the remark this answers is not in the document]
Comment (runame) [reply matrix_functions.py:266]: We will reuse this for the adaptive amortized computation frequency implementation (and also mock it in that context).

File: distributed_shampoo/utils/tests/shampoo_preconditioner_list_test.py [hunk 2]
```diff
@@ -338,13 +338,21 @@ def test_raise_nan_in_factor_matrix(self) -> None:
         def test_raise_nan_and_inf_in_inv_factor_matrix_amortized_computation(
             self,
         ) -> None:
-            for invalid_value in (torch.nan, torch.inf):
+            invalid_values = (
+                (torch.tensor([torch.nan]), torch.tensor([torch.inf]))
+                if self._amortized_computation_function() == "matrix_inverse_root"
+                else (  # matrix_eigendecomposition returns a tuple.
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/utils/tests/shampoo_preconditioner_list_test.py:344]: I was also considering this, maybe a good solution would be a `matrix_eigenvectors` function that has the same signature as `matrix_eigendecomposition` and returns `matrix_eigendecomposition(...)[1]`?

File: matrix_functions_types.py [hunk 3]
```diff
@@ -19,25 +19,58 @@ class MatrixFunctionConfig(AbstractDataclass):
...
+    Attributes:
+        max_iterations (int): The maximum number of iterations to perform. (Default: 1)
+        tolerance (float): The tolerance for determining convergence in terms of the norm of the off-diagonal elements of the eigenvalue estimate.
+            (Default: 0.01)
+
+    """
+
+    max_iterations: int = 1
+    tolerance: float = 0.01
+
```
[the remark this answers is not in the document]
Comment (runame) [reply matrix_functions_types.py:69]: We implemented or at least discussed implementing this for the deleted `matrix_eigenvectors` function when we first implemented it, but decided against it. I'm not exactly sure why anymore, but can think about it again tomorrow.

[the remark this answers is not in the document]
Comment (runame) [reply matrix_functions_types.py:71]: > The other possibility could be as simple as we forgot field(init=False).

You definitely suggested this last time as well, but I can't remember at all why we didn't go for it in the end. So I just went ahead and implemented your suggestion.

File: distributed_shampoo/utils/shampoo_preconditioner_list.py [hunk 4]
```diff
@@ -936,8 +940,8 @@ def _create_kronecker_factors_list(
             factor_matrix_indices=kronecker_factors_state.factor_matrix_indices,
         )
 
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/utils/shampoo_preconditioner_list.py:943]: Weird, shouldn't it cause an error _without_ the `staticmethod` decorator?

[the remark this answers is not in the document]
Comment (runame) [reply distributed_shampoo/utils/shampoo_preconditioner_list.py:943]: Discussed offline, also added `staticmethod` decorator to `abstractmethod`.

File: matrix_functions_types.py [hunk 5]
```diff
@@ -19,25 +19,62 @@ class MatrixFunctionConfig(AbstractDataclass):
...
+            )
+
+
+DefaultEigendecompositionConfig = EighEigendecompositionConfig()
+
+
+@dataclass(kw_only=True)
+class QREigendecompositionConfig(EigendecompositionConfig):
+    """Configuration for eigenvalue decomposition via QR algorithm.
+
```
Comment (hjmshi) [inline matrix_functions_types.py:55]: Perhaps this is already known or obvious, but we can probably mention here the observation that checking the diagonality of $$B$$ is equivalent to checking how good of an approximation we have of $$A$$. 

Specifically, if $$\tilde{A}$$ is an approximation of $$A$$ using an approximate orthogonal basis $$\tilde{Q}$$ and eigenvalues $$\tilde{\Lambda} = diag(\tilde{Q}^T A \tilde{Q})$$, then:
$$\|A - \tilde{A}\|_F = \|A - \tilde{Q} \tilde{\Lambda} \tilde{Q}^T\|_F = \|\tilde{Q}^T A \tilde{Q} - \tilde{\Lambda}\|_F = \|B - diag(B)\|_F.$$
The RHS of the condition also winds up being equivalent, i.e.,
$$\|B\|_F = \|\tilde{Q}^T A \tilde{Q}\|_F = \|A\|_F$$.

One can also see this by simply noting that orthogonal transformations preserve the eigenvalues of the matrix, and the Frobenius norm gives you the 2-norm of the eigenvalue vector.

This all holds since $$\|Q A\|_F^2 = Tr((QA)^T(QA)) = Tr(A^TQ^T Q A) = Tr(A^T A) = \|A\|_F^2$$.

cc: @vinaysrao

File: matrix_functions.py [hunk 6]
```diff
@@ -206,13 +257,88 @@ def matrix_eigenvalue_decomposition(
...
+    Returns:
+        bool: Whether the criterion is below or equal to the tolerance
+
+    """
+    squared_approximate_eigenvalues = torch.square_(Q.T @ A @ Q)
+    diagonal_summed = squared_approximate_eigenvalues.diag().sum()
+    off_diagonal_summed = squared_approximate_eigenvalues.fill_diagonal_(0.0).sum()
+    norm = torch.sqrt(diagonal_summed + off_diagonal_summed)
+    off_diagonal_norm = torch.sqrt(off_diagonal_summed)
+    return bool(off_diagonal_norm <= tolerance * norm)
```
Comment (hjmshi) [inline matrix_functions.py:283]: Wondering if it may be faster to just compute the full norm without filling the diagonal and subtracting? 
```suggestion
    approximate_eigenvalues = Q.T @ A @ Q
    total_norm = torch.linalg.norm(approximate_eigenvalues)
    diagonal_norm = torch.linalg.norm(A.diag())
    off_diagonal_norm = torch.sqrt(A_norm ** 2 - diagonal_norm ** 2)
    return bool(off_diagonal_norm <= tolerance * total_norm)
```

File: matrix_functions.py [hunk 7]
```diff
@@ -206,13 +257,88 @@ def matrix_eigenvalue_decomposition(
...
+    Args:
+        A (Tensor): The symmetric input matrix.
+        Q (Tensor): An estimate of the eigenvectors of A.
+        tolerance (float): The tolerance for the criterion.
+
+    Returns:
+        bool: Whether the criterion is below or equal to the tolerance
+
+    """
+    squared_approximate_eigenvalues = torch.square_(Q.T @ A @ Q)
```
Comment (hjmshi) [inline matrix_functions.py:278]: Do we want to return `Q.T @ A @ Q` as well (or its diagonal)? We should cache this value and use it if the test succeeds.

We can also use the einsum trick used below for computing `Q.T @ A @ Q` if it is faster. However, @shintaro-iwasaki thinks that `einsum` may not be performant.

File: matrix_functions.py [hunk 8]
```diff
@@ -206,13 +257,88 @@ def matrix_eigenvalue_decomposition(
...
+        and not _approximate_eigenvalues_criterion_below_or_equal_tolerance(
+            A, Q, tolerance
+        )
+    ):
+        power_iteration = A @ Q
+        Q = torch.linalg.qr(power_iteration).Q
+        iteration += 1
+
+    # Ensure consistent ordering of estimated eigenvalues and eigenvectors.
+    estimated_eigenvalues, indices = torch.einsum("ij, ik, kj -> j", Q, A, Q).sort()
```
Comment (hjmshi) [inline matrix_functions.py:329]: What is the purpose of sorting in this case? Do we need this?

**Review by hjmshi — COMMENTED** [review]
Thanks @runame for putting this together! Left some comments below. Haven't looked at the tests yet, but wanted to get this to you sooner. :)

File: matrix_functions_types.py (hunk 5 above)
Comment (runame) [reply matrix_functions_types.py:55]: Agreed that spelling out the motivation here is helpful, added it in [this commit](https://github.com/facebookresearch/optimizers/pull/88/commits/fead22105e5214e5d35ea3a785409a388b15604d).

File: matrix_functions.py (hunk 6 above)
Comment (runame) [reply matrix_functions.py:288]: It seems to be roughly the same, but your suggestion is slightly faster in most settings. I did some additional benchmarking and the speed-ups for the largest matrices seems to be due to the use of `torch.linalg.norm`.

Using `einsum` is a lot slower for smaller matrices and faster for only the largest tested matrix size. Full results:

Screenshot 2025-03-18 at 13 52 01

File: matrix_functions.py (hunk 7 above)
Comment (runame) [reply matrix_functions.py]: Added [this](https://github.com/facebookresearch/optimizers/pull/88/commits/26a4ccc515b24175c19b114877011d01b94d0788).

Regarding `einsum` inside of the criterion see my comment above. For extracting the diagonal the slow-down for small matrices seems to be less severe, but the speed-up for large matrices more significant, so I used `einsum` here.

File: matrix_functions.py (hunk 8 above)
Comment (runame) [reply matrix_functions.py]: This won't affect eigendecomposed Shampoo since all that matters is that the eigenvectors and eigenvalues are sorted consistently. However, for eigenvalue-corrected Shampoo we don't use the eigenvalues and accumulate the squared gradient in a transformed basis. So if the ordering of the columns of the transformation matrix (the eigenvectors) changes every iteration (extreme case), this _could_ affect things.

Comment (runame) [reply matrix_functions.py]: Yes, I also noticed that we currently _don't_ sort the eigenvalues for diagonal matrices, which is inconsistent. I just fixed that and also decreased the numerical tolerance of the tests by increasing the precision of the hard-coded target values.

Comment (runame) [reply matrix_functions.py]: OK, this broke the preconditioner list tests because they depend on the lack of sorting for diagonal matrices.

File: matrix_functions.py [hunk 9]
```diff
@@ -206,13 +258,109 @@ def matrix_eigenvalue_decomposition(
     return L.to(device=current_device), Q.to(device=current_device)
 
 
+class Criterion(NamedTuple):
```
[the remark this answers is not in the document]
Comment (runame) [reply matrix_functions.py:261]: We will reuse the function `_approximate_eigenvalues_criterion_below_or_equal_tolerance`, but the `Criterion` is only used for this function for now. We could 1) move the `Criterion` definition inside of the function scope, and/or 2) change its name to `ApproximateEigenvaluesCriterion`. Any preference?

File: matrix_functions.py [hunk 10]
```diff
@@ -321,16 +332,23 @@ def _qr_algorithm(
...
         iteration += 1
 
     # Ensure consistent ordering of estimated eigenvalues and eigenvectors.
-    estimated_eigenvalues, indices = torch.einsum("ij, ik, kj -> j", Q, A, Q).sort()
+    estimated_eigenvalues = (
+        criterion.approximate_eigenvalues
+        if iteration == 0  # Re-use approximate eigenvalues if iterations were skipped.
+        else torch.einsum("ij, ik, kj -> j", Q, A, Q)
+    )
+    estimated_eigenvalues, indices = estimated_eigenvalues.sort()
```
[the remark this answers is not in the document]
Comment (runame) [reply matrix_functions.py]: I think I prefer the current code structure, mainly because 1) we only need one call to the criterion function, which highlights that there is no fundamental difference between the initial check which might result in skipping the iterations and the check for convergence, and 2) a while-loop is more in spirit with the logic of the algorithm than of a for-loop with `break` condition.

Can you explain more what about the `if iteration == 0` check looks weird to you?
